MovingBlocks/Terasology
# Refetch controllers by clicking on menu button

### Contains
This pull request begin with PR #3705 as a base for changes.

* A "Refresh Controllers" button was added to main menu that partially
solves "recognize disconnect and reconnect" feature in issue/#2125. When
this button is pressed, the list of controllers are re-fetched to be
recognized by the system. This partially solves the problem of having
to close and re-open the game every time a controller is created.

* Also, when this button is pressed, the "Controller Settings" screen is
disposed from the assets. This is done to solve to force the system to re-create
the "Controller Settings" screen in case it had already loaded in the assets
before the "Refresh Controllers" was pressed. This makes the system show the controllers
in the "Controller Settings" screen after re-fetching the controllers. That is why it is related to issue/#3648.

Note that: the re-fetching of controllers was implemented and tested only for windows 10 and Xbox 360 controller. Implementing and testing needs to be done for other operating systems and controllers.

### How to test
Run the game as usual without connecting a controller. You can begin a new game if you want. Then go back to the main menu. Connect a controller. Click on "Refresh Controllers". Then, go to "Settings" -> "Controller Settings" if you want. You should see that the controller options are shown. Then you can go back to the main menu and continue playing using the controller.

**Note that**: when resuming the game in the manner described above, sometimes you need to move the mouse a little bit on the screen and see that player is rotating. Then, begin playing with the controller. I think that it has to do with focus on screen or something like that.

It was a little bit hard to write unit tests (two unit tests only were written) to test this feature because it was hard to hard to simulate plugging and unplugging a controller.

### Outstanding before merging

- [ ] Need to implement and test for other operating systems (e.g. linux and macOS) and other types of controllers.
- [ ] Need more unit tests.
- [ ] Need to have a way to not need the "Refresh Controllers" totally. Maybe JNI is a good option. Or move "Refresh Controllers" button to be in "Controller Settings" menu instead.
